File: algorithms/code/leetcode/lc104_maximum_depth_of_binary_tree/lc104_maximum_depth_of_binary_tree.py
# https://leetcode.com/problems/maximum-depth-of-binary-tree/

from __future__ import annotations

import unittest


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: TreeNode | None = None,
        right: TreeNode | None = None,
    ):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxDepth(self, root: TreeNode | None) -> int:
        if root is None:
            return 0
        left = self.maxDepth(root.left)
        right = self.maxDepth(root.right)
        return 1 + max(left, right)


class TestCode(unittest.TestCase):
    def test_example(self) -> None:
        n0 = TreeNode(0)
        n1 = TreeNode(1)
        n2 = TreeNode(2)
        n3 = TreeNode(3)
        n0.left = n1
        n0.right = n2
        n1.left = None
        n1.right = n3
        n2.left = None
        n2.right = None
        n3.left = None
        n3.right = None
        assert Solution().maxDepth(n0) == 3

File: algorithms/code/leetcode/lc104_maximum_depth_of_binary_tree/lc104_maximum_depth_of_binary_tree.rs
// https://leetcode.com/problems/maximum-depth-of-binary-tree/
// #easy

use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    pub fn new(
        val: i32,
        left: Option<Rc<RefCell<TreeNode>>>,
        right: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<Self>>> {
        Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
    }
}

pub struct Solution;

impl Solution {
    pub fn max_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        match root {
            None => 0,
            Some(root) => {
                let root = root.borrow();
                let left_depth = Self::max_depth(root.left.clone());
                let right_depth = Self::max_depth(root.right.clone());
                1 + left_depth.max(right_depth)
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        let n3 = TreeNode::new(3, None, None);
        let n1 = TreeNode::new(1, None, n3);
        let n2 = TreeNode::new(2, None, None);
        let n0 = TreeNode::new(0, n1, n2);
        assert_eq!(Solution::max_depth(n0), 3);
    }
}
